core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1513,359 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
+	 * entries fit into the temporary buffer.
+	 */
+	if (CFG_RPMB_FS_CACHE_SIZE)
+		num_elems_read = MIN(CFG_RPMB_FS_CACHE_SIZE,
+				     CFG_RPMB_FS_RD_ENTRIES);
+	else
+		num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
 
-		for (i = 0; i < CFG_RPMB_FS_RD_ENTRIES; i++) {
+	/*
+	 * Allocate memory for the FAT FS entries to read in.
+	 */
+	fe = calloc(num_elems_read, sizeof(struct rpmb_fat_entry));
+	if (!fe) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
 
-			FMSG("flags 0x%x, size %d, address 0x%x, filename '%s'",
-				fat_entries[i].flags,
-				fat_entries[i].data_size,
-				fat_entries[i].start_address,
-				fat_entries[i].filename);
+	res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID, fat_address, (uint8_t *)fe,
+			    num_elems_read * sizeof(*fe), NULL, NULL);
+	if (res)
+		goto out;
 
-			if ((fat_entries[i].flags & FILE_IS_LAST_ENTRY) != 0) {
-				last_entry_found = true;
-				break;
+	fat_entry_dir->rpmb_fat_entry_buf = fe;
+
+	/*
+	 * We use this variable when getting next entries from the buffer/cache
+	 * to see whether we have to read in more entries from storage.
+	 */
+	fat_entry_dir->num_buffered = num_elems_read;
+
+out:
+	if (res && fat_entry_dir)
+		fat_entry_dir_free();
+
+	return res;
+}
+
+/**
+ * fat_entry_dir_deinit: If caching is enabled, free the temporary buffer for
+ * FAT FS entries in case the cache was too small. Keep the elements in the
+ * cache. Reset the counter variables to start the next traversal from fresh
+ * from the first cached entry. If caching is disabled, just free the
+ * temporary buffer by calling fat_entry_dir_free and return.
+ */
+static void fat_entry_dir_deinit(void)
+{
+	struct rpmb_fat_entry *fe = NULL;
+
+	if (!fat_entry_dir)
+		return;
+
+	if (!CFG_RPMB_FS_CACHE_SIZE) {
+		fat_entry_dir_free();
+		return;
+	}
+
+	fe = fat_entry_dir->rpmb_fat_entry_buf;
+	fat_entry_dir->idx_curr = 0;
+	fat_entry_dir->num_total_read = 0;
+	fat_entry_dir->last_reached = false;
+
+	/*
+	 * If the cache is full and we used our temporary buffer, only keep
+	 * cached entries, shrink down to cache size.
+	 */
+	if (fat_entry_dir->num_buffered > CFG_RPMB_FS_CACHE_SIZE) {
+		fat_entry_dir->num_buffered = CFG_RPMB_FS_CACHE_SIZE;
+		/* We reduce the size of the alloc, so this should be good. */
+		fe = realloc(fe, fat_entry_dir->num_buffered * sizeof(*fe));
+
+		/*
+		 * In case realloc fails, we are on the safe side if we destroy
+		 * the whole structure. Upon the next init, the cache has to be
+		 * re-established, but this case should not happen in practice.
+		 */
+		if (!fe)
+			fat_entry_dir_free();
+		else
+			fat_entry_dir->rpmb_fat_entry_buf = fe;
+	}
+}
+
+/**
+ * fat_entry_dir_update: Updates a persisted FAT FS entry in the cache.
+ * This function updates the FAT entry fat_entry that was written to address
+ * fat_address onto RPMB storage in the cache.
+ * Returns TEE_SUCCESS on success.
+ */
+static TEE_Result fat_entry_dir_update(struct rpmb_fat_entry *fat_entry,
```

Comment:
Maybe rename to `fat_cache_update()`.
Suggest attribute `__maybe_unused` as never called when `CFG_RPMB_FS_CACHE_SIZE == 0`.